Android Signer Application (#868)
nostr-protocol/nips: 100.md

# NIP-100

## Android Signer Application

`draft` `optional`

> This NIP describes a method for 2-way communication between a android signer and any Nostr client on Android. The Android signer is an Android Application and the Client can be a Web Client or an Android Application.

Reviewer: ```suggestion
This NIP describes a method for 2-way communication between an Android signer and any Nostr client on Android. The Android signer is an Android Application and the client can be a web client or an Android application.
```

# Usage for Android applications

The Android signer uses Intents and Content Resolvers to communicate between applications.

> To be able to use The Android signer in your application you should this to your AndroidManifest.xml:

Reviewer: ```suggestion
To be able to use the Android signer in your application you should add this to your AndroidManifest.xml:
```

```xml
<queries>
  <intent>
    <action android:name="android.intent.action.VIEW" />
    <category android:name="android.intent.category.BROWSABLE" />
    <data android:scheme="nostrsigner" />
  </intent>
</queries>
```

Then you can use this function to check if there's a signer application installed:

```kotlin
fun isExternalSignerInstalled(context: Context): Boolean {
  val intent =
    Intent().apply {
      action = Intent.ACTION_VIEW
      data = Uri.parse("nostrsigner:")
    }
  val infos = context.packageManager.queryIntentActivities(intent, 0)
  return infos.size > 0
}
```

## Using Intents

> To get the result back from the Signer Appication you should use registerForActivityResult or rememberLauncherForActivityResult in Kotlin. If you are using another framework check the documentation of your framework or a third party library to get the result.

Reviewer: ```suggestion
To get the result back from the Signer Application you should use `registerForActivityResult` or `rememberLauncherForActivityResult` in Kotlin. If you are using another framework check the documentation of your framework or a third party library to get the result.
```

> To get the result back from the Signer Application you should use `registerForActivityResult` or `rememberLauncherForActivityResult` in Kotlin. If you are using another framework check the documentation of your framework or a third party library to get the result.

Reviewer: I personally would appreciate a simple example for using `rememberLauncherForActivityResult`. It would have made things easier for me :D

```kotlin
val launcher = rememberLauncherForActivityResult(contract = ActivityResultContracts) { activityResult ->
  val signature = activityResult.data?.getStringExtra("signature")
  // Do something with signature ...
}

launcher.launch(input = intent, options = ActivityOptionsCompat.makeBasic())
```

      // the id you sent
      val id = intent.data?.getStringExtra("id")
      ```        

- **decrypt_zap_event**
  - params:

    ```kotlin
    val intent = Intent(Intent.ACTION_VIEW, Uri.parse("nostrsigner:$eventJson"))
    intent.`package` = "com.example.signer"
    intent.putExtra("type", "decrypt_zap_event")
    // to control the result in your application in case you are not waiting the result before sending another intent
    intent.putExtra("id", "some_id")
    // Send the current logged in user npub
    intent.putExtra("current_user", account.keyPair.pubKey.toNpub())
    context.startActivity(intent)
    ```
  - result:
    - If the user approved intent it will return the **signature** and **id** fields

      ```kotlin
      val eventJson = intent.data?.getStringExtra("signature")
      // the id you sent
      val id = intent.data?.getStringExtra("id")
      ```         

## Using Content Resolver

To get the result back from Signer Application you should use contentResolver.query in Kotlin. If you are using another framework check the documentation of your framework or a third party library to get the result.

> If the user did not check the remember my choice option, the npub is not in Signer Application or the signer type is not recognized the contentResolver will return null

Reviewer: ```suggestion
If the user did not check the "remember my choice" option, the npub is not in Signer Application or the signer type is not recognized the `contentResolver` will return null
```

For the SIGN_EVENT type Signer Application returns two columns "signature" and "event". The column event is the signed event json

For the other types Signer Application returns the column "signature"

> If the user chose to always reject the event signer application will return the column "rejected" and you should not open signer application

Reviewer: ```suggestion
If the user chose to always reject the event, signer application will return the column "rejected" and you should not open signer application
```

        }
      ```    

- **nip44_decrypt**
  - params:

    ```kotlin
    val result = context.contentResolver.query(
        Uri.parse("content://com.example.signer.NIP44_DECRYPT"),
        listOf("$encryptedText", "${hex_pub_key}", "${logged_in_user_npub}"),
        null,
        null,
        null
    )
    ```
  - result:
    - Will return the **signature** column

      ```kotlin
        if (result == null) return

        if (result.moveToFirst()) {
            val index = it.getColumnIndex("signature")
            val encryptedText = it.getString(index)
        }
      ```      

- **decrypt_zap_event**
  - params:

    ```kotlin
    val result = context.contentResolver.query(
        Uri.parse("content://com.example.signer.DECRYPT_ZAP_EVENT"),
        listOf("$eventJson", "", "${logged_in_user_npub}"),
        null,
        null,
        null
    )
    ```
  - result:
    - Will return the **signature** column

      ```kotlin
        if (result == null) return

        if (result.moveToFirst()) {
            val index = it.getColumnIndex("signature")
            val eventJson = it.getString(index)
        }
      ```    

# Usage for Web Applications

> Since web applications can't receive a result from the intent you should add a modal to paste the signature or the event json or create a callback url.

Reviewer: ```suggestion
Since web applications can't receive a result from the intent, you should add a modal to paste the signature or the event json or create a callback url.
```

> If you send the callback url parameter Signer Application will send the result to the url.

Reviewer: ```suggestion
If you send the callback url parameter, Signer Application will send the result to the url.
```

> If you don't send a callback url Signer Application will copy the result to the clipboard.

Reviewer: ```suggestion
If you don't send a callback url, Signer Application will copy the result to the clipboard.
```

> You can configure the returnType to be **signature** or **event**.

Reviewer: ```suggestion
You can configure the `returnType` to be **signature** or **event**.
```

> Android intents and browsers url has limitations, so if you are using the returnType of **event** consider using the parameter **compressionType=gzip** that will return "Signer1" + Base 64 gzip encoded event json

Reviewer: ```suggestion
Android intents and browser urls have limitations, so if you are using the `returnType` of **event** consider using the parameter **compressionType=gzip** that will return "Signer1" + Base64 gzip encoded event json
```

    window.href = intent;
    ```     

- **decrypt_zap_event**
  - params:

    ```js
    const intent = `intent:${eventJson}#Intent;scheme=nostrsigner;S.compressionType=none;S.returnType=signature;S.type=decrypt_zap_event;S.callbackUrl=https://example.com/?event=;end`;

    window.href = intent;
    ```          

## Example

```js
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>
<body>
    <h1>Test</h1>
       
    <script>
        window.onload = function() {
            var url = new URL(window.location.href);
            var params = url.searchParams;
            if (params) {
                var param1 = params.get("event");
                if (param1) alert(param1)
            }
            let json = {
                kind: 1,
                content: "test"
            }
            let encodedJson = encodeURIComponent(JSON.stringify(json))
            var newAnchor = document.createElement("a");

>             newAnchor.href = `intent:${encodedJson}#Intent;scheme=nostrsigner;S.compressionType=none;S.returnType=signature;S.type=sign_event;S.callbackUrl=https://example.com/?event=;end`;

Reviewer 2: This is notably an Android-centric URI. For example, iOS doesn't have intents. Does this mean that web applications need to start being platform aware? If so, I'm not sure if this is a good choice. Can we achieve the same result in a platform-agnostic way?

I've been thinking through how to build an iOS signer application, and I feel like we can shape this spec to be platform agnostic (while still allowing flexibility for platform-specific first-class features be used to make the UX better).

Something like
```js
newAnchor.href = `nostrsigner:${encodedJson}?compressionType=none&S.returnType=signature;S.type=sign_event&callbackUrl=https://example.com/?event=;end
```

Let me know if I've misunderstood anything.

Author: We can change to do the way you did so it is platform-agnostic.
